### ISSUE TITLE
Parallelize file processing

### ISSUE THREAD

**Issue by dmnks** [rpm-software-management/rpm#1185]
My preliminary testing shows a promising reduction of build time (with the kernel package) when running the loop inside `build/files.c:processBinaryFiles()` in parallel (more specifically, the `processPackageFiles()` call), but I still have to analyze and understand the code better, to see if it really is safe to do and nothing breaks as a result (with respect to race conditions etc.)

I'll attach a WIP PR once it's in a readable form, to provide more context.

**Comment by pmatilai** [comment]
Based on a quick look, much of it does in fact appear thread-safe. There are grues in the darker corners though :eyes: 

One certain problem is the global macro space, readFilesManifest() does push/pop on %license which is probably harmless, but further down the call chain there's readManifest() and specExpand() which do push/pop on current file name and line number, which is not. So at the very least the files manifest parsing would have to be lifted to a pre-parallel section. I'd be nothing short of shocked if this was the only gotcha there is, as processPackageFiles() ends up covering a scary amount of code.

**Comment by dmnks** [comment]
Thanks for the feedback.

I also noticed `readFilesManifest()` as being one potential source of problems here. Indeed, with a dirty patch that I'm currently testing out, I got a few random crashes (possibly due to that).

However, in those cases where it worked, the build time of a kernel on remote 16-core system with a F31 went down from ~7m to ~3m (just the build phase, i.e. `rpmbuild -bb --short-circuit --noclean`).

That makes it worthy a more thorough look at least, so I'll see what I can do :) There's nothing to lose; I'll learn something about the codebase either way.

**Comment by ignatenkobrain** [comment]
@dmnks this sounds very nice, is there some PR to follow?

**Comment by dmnks** [comment]
Good to know! No PR yet, but I'll create one soon-ish and reference this issue in it.

**Comment by dmnks** [comment]
So I did some more testing and it turns out rpmbuild only spends a tiny fraction of time in the `processPackageFiles()` function; dependency generators (kmod.prov in particular) are a much bigger bottleneck but also vastly trickier to parallelize.

The speed improvements that I observed a while ago in https://github.com/rpm-software-management/rpm/issues/1185#issuecomment-617147116 were most likely just a false positive due to some internal data inconsistency caused by the "naive" patch that I tested with.

With that said, I'm closing this ticket as there's no point in putting much effort in this. OTOH, parallel execution of dependency generators (`rpmfcGenerateDepends()`) is something where real improvements could be made (in the case of kernel, the cumulative runtime of these is in the order of hunderds of seconds). But that would be tracked in a separate issue.
